### ISSUE TITLE
tags aren't arranged alphabetically as the standard require

### ISSUE THREAD

**Issue by riksoft** [fontello/fontello#607]
Hi guys, 
I've got this warning in Firefox:

**downloadable font: Layout: tags aren't arranged alphabetically. (font-family: "websymbol" style:normal weight:normal stretch:normal src index:1) source:....**

...where Websymbols is a webfont generated by Fontello.

Asking info about such unconprehensible message to Mozilla they told me that the OpenType specification requires several lists of tags (scripts, language systems, features) to be ordered alphabetically in the GSUB and GPOS tables and the Fontello tool doesn't.

In this page
https://www.microsoft.com/typography/otspec/chapter2.htm
searching the word "alphabetically" you can find the set of stuff that should be alphabetically ordered to be standard compliant.

**Comment by puzrin** [comment]
I never seen such warnign before.

1. Could you provide `config.json` from minimal sample (remove icons one-by-one until problem exists in demo page)?
2. What is you FF / OS version?

**Comment by riksoft** [comment]
You won't see such warning in the standard release of Firefox (v 53), but I'm on the Aurora channel of Firefox (Firefox Developer Edition 54.0a2 (2017-04-18) (64-bit)) and they have strengthen the validator. This will be soon on the standard version too.

OS: Linux Mint 18.1 Serena 64 bit,
DE: Mate

The warning is on a .woff that is only 7K containing 23 symbols all from fontawesome except 1 from  fontelico:
"uid": "f0c301ac841dafc38d8eb1b933fc73e5",
"css": "spin",
"code": 59393,
"src": "fontelico"

If you think it's important I'll try to reduce such list.

**Comment by puzrin** [comment]
1. If possible - reduce it, yes. If not possible to reduce - attach config as is. Finally, i need config to reproduce bug on my side.
2. Try to enable/disable hinting. Will it change something?

**Comment by riksoft** [comment]
I did some tests and I've reduced it to the bare minimum: I've chosen only the 1st symbol from fontawesome, downloaded the package and launched "demo.htm". Same thing choosing only the 1st symbol of fontelico. Both get warnings.

The 2 packages of test are together only 51 kB so I can attach them directly to this post.
[fontello-3c78c192.zip](https://github.com/fontello/fontello/files/1031250/fontello-3c78c192.zip)
[fontello-839d08eb.zip](https://github.com/fontello/fontello/files/1031251/fontello-839d08eb.zip)

**To recap:**
- open Firefox 54 (Developer Edition... ore wait some weeks for the standard version)
- CTRL + SHIFT + k to open the console
- Then open "demo.html" that is always present in the fontello zip package
and you'll se the following warning:
`downloadable font: Layout: tags aren't arranged alphabetically. (font-family: "fontello" style:normal weight:normal stretch:normal src index:1) source: file:///home/.../fontello-3c78c192/font/fontello.woff?60804650`

**Comment by puzrin** [comment]
1. I've reproduced this problem. It seems to happen only with enabled hinting. Can you confirm?
2. Fontello still use very old ttfautohint v1.1, but i'm not sure this issue was ever fixed.
3. This problem is not fatal, just warning, but everything still works.

@lemzwerg can you comment? How to reproduce:

- use FF 54 (aurora) from http://ftp.mozilla.org/pub/mozilla.org/firefox/nightly/latest-mozilla-aurora/
- enable hinting in fontello and download any font (ttfa 1.1 will be used)
- open demo page and see warnig in console about bad tags order

**Comment by lemzwerg** [comment]
This problem is not related to ttfautohint.  I've downloaded a font from fontello with and without hinting, and testing the font with `ots-sanitize` (from https://github.com/khaledhosny/ots) I get the the layout warning for both fonts.

**Comment by puzrin** [comment]
@lemzwerg seems you are right. [ots-sanitize](https://github.com/khaledhosny/ots) from master report this warning for all cases. Thank you very much for the link!

@sabberworm could you take a look at https://github.com/fontello/svg2ttf/blob/master/lib/ttf/tables/gsub.js? Git [history](https://github.com/fontello/svg2ttf/commits/master/lib/ttf/tables/gsub.js) says, you are author of it :)

**Comment by puzrin** [comment]
Seems to be fixed in my simple checks. If any problem still exists - let be know.

@riksoft , @lemzwerg thanks again for your help!

**Comment by sabberworm** [comment]
@puzrin I’ve seen the warning before, it just didn’t occur to me that it was stemming from my code… Thanks for the fix.

**Comment by puzrin** [comment]
@sabberworm `ots-sanitize` is more informative. It said about problem with script tags order, and i just searched word "script" in sources :). Probably, this tool should be added to CI tests.

**Comment by riksoft** [comment]
Sorry for my huge delay of reply: I confirm now it's all OK! :)
Thanks Puzrin!
And by the way, thanks a lot for such a great website!
